### PR TITLE
Callback to track when the llvm lock is acquired and released

### PR DESCRIPTION
**Pull request by sklam**
as titled.

### CODE REVIEW

File: llvmlite/binding/ffi.py [hunk 1]
```diff
@@ -37,6 +37,46 @@ def _make_opaque_ref(name):
 LLVMSectionIteratorRef = _make_opaque_ref("LLVMSectionIterator")
 
 
+class _LLVMLock:
+    """A Lock to guarantee thread-safety for LLVM C-API.
```
Comment (stuartarchibald) [inline llvmlite/binding/ffi.py:41]:
```suggestion
    """A Lock to guarantee thread-safety for the LLVM C-API.
```

File: llvmlite/binding/ffi.py [hunk 2]
```diff
@@ -37,6 +37,46 @@ def _make_opaque_ref(name):
...
+    def __init__(self):
+        # The reentrant lock is needed for callbacks that re-enter
+        # the Python interpreter.
+        self._lock = threading.RLock()
+        self._cblist = []
+
+    def register(self, acq_fn, rel_fn):
+        """Register callbacks that is invoked immediately after the lock is
+        acquired (``acq_fn()``) and immediately before the lock is released
+        (``rel_fn()``).
```
Comment (stuartarchibald) [inline llvmlite/binding/ffi.py:58]:
```suggestion
        """Register callbacks that are invoked immediately after the lock is
        acquired (``acq_fn()``) and immediately before the lock is released
        (``rel_fn()``).
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, looks good. Two minor things to resolve then good to merge.

**Comment by sklam** [comment]
Thanks for reviewing. All comments are addressed.

**Comment by stuartarchibald** [comment]
> Thanks for reviewing. All comments are addressed.

Thanks @sklam. Do you think additional changes to this PR may be required in order to resolve this: https://github.com/numba/numba/pull/6636#issuecomment-760168944 ? If not, will mark for merge.

**Comment by sklam** [comment]
https://github.com/numba/numba/pull/6636#issuecomment-760168944 is resolved. No need for changes here.

**Comment by stuartarchibald** [comment]
> [numba/numba#6636 (comment)](https://github.com/numba/numba/pull/6636#issuecomment-760168944) is resolved. No need for changes here.

Thanks for confirming.

**Pull request merged**